Add NaK Fluid properties

GiudGiud opened the pull request:
closes #24450

GiudGiud committed:
Add fluid properties for NaK eutectic

GiudGiud committed:
Add unit test for NaK
refs #24450

GiudGiud committed:
Add range to correlations from NaK handbook
refs #24450

GiudGiud committed:
Add missing routines for use of NaK in subchannel
refs #24450

GiudGiud committed:
Add the solution invalidity interface to fluid properties for checking ranges

GiudGiud committed:
Uniformize the variables in function definitions
Fixup units
refs #24450

GiudGiud committed:
Add documentation for NaK fluid properties

GiudGiud commented on modules/fluid_properties/src/userobjects/NaKFluidProperties.C:
```diff
@@ -0,0 +1,270 @@
+  params.addClassDescription("Fluid properties for NaK");
+  params.addRequiredParam<Real>(
+      "weight_fraction_K",
+      "Weight fraction of potassium in NaK. Only eutectic is implemented (X_K = 0.778)");
```
this seems unnecessary but it matters a lot so I want people to realize that when they make their input file

lindsayad replied:
Would you just want to name this as `EutecticNaKFluidProperties` and then not have this parameter that will likely force users to get an error at least once? It's counter-intuitive to require an input that can only be one thing

GiudGiud replied:
it s required though so they will specify it

I like NaK because it goes next to NaCl in the index. Prefer eutectic?

lindsayad replied:
I do, but maybe another reviewer will feel otherwise. @joshuahansel ?

lindsayad replied:
@tanoret ?

lindsayad replied:
> if they know what the eutectic is

That and know to specify it to exactly three significant figures, so yea they would likely have to read the parameter description first ... or just copy it from example input files which is probably what they will be doing. I suppose that this is ok

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1543974/) on 52012ae wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24452/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24452/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 395ab6da47fe9a63aab358a9ca122e2fd6dd943d`

GiudGiud committed:
Add yaml unit test files to gitignore, follow up on #24167

GiudGiud commented on .gitignore:
```diff
@@ -294,3 +294,30 @@ share/
 /modules/reactor/reactor.yaml
 /modules/rdg/rdg.yaml
 /modules/fsi/fsi.yaml
+/unit/moose-unit.yaml
```
@milljm any issue with that?

lindsayad commented on modules/fluid_properties/include/userobjects/NaKFluidProperties.h:
```diff
@@ -0,0 +1,86 @@
+/**
+ * NaK fluid properties as a function of pressure (Pa) and temperature (K).
+ *
+ * Origin of fluid properties:
+ * 1) density; thermal diffusivity; dynamic viscosity; specific heat capacity
```
I don't know what this means. Does it mean this object computes these properties or requires them as inputs?

GiudGiud replied:
no it means these fluids properties comes from this book

GiudGiud replied:
if we have more properties in the future they can come from the same book or another book, just list it there in the header

lindsayad commented on modules/fluid_properties/include/userobjects/NaKFluidProperties.h:
```diff
@@ -0,0 +1,86 @@
+
+  Real T_from_p_rho(Real pressure, Real density) const;
+
+  virtual Real rho_from_p_T(Real pressure, Real temperature) const override;
+
+  virtual void rho_from_p_T(
```
Let's group overloads with the same name without spacing
```suggestion
  virtual Real rho_from_p_T(Real pressure, Real temperature) const override;
  virtual void rho_from_p_T(
```

GiudGiud replied:
addressed that

lindsayad commented on modules/fluid_properties/src/userobjects/NaKFluidProperties.C:
```diff
@@ -0,0 +1,270 @@
+{
+  auto lambda = [&](Real p, Real current_T, Real & new_h, Real & dh_dp, Real & dh_dT)
+  { h_from_p_T(p, current_T, new_h, dh_dp, dh_dT); };
+  Real T = FluidPropertiesUtils::NewtonSolve(
+               pressure, enthalpy, _T_initial_guess, _tolerance, lambda, name() + "::T_from_p_h")
+               .first;
```
It makes me feel better to see this getting used

GiudGiud replied:
We could avoid it here. It s a third order polynomial there's a method to solve them.

lindsayad commented on modules/fluid_properties/src/userobjects/NaKFluidProperties.C:
```diff
@@ -0,0 +1,270 @@
+  const Real v_k = 1. / (0.8415 - 2.172e-4 * Tc - 2.7e-8 * Tc * Tc + 4.77e-12 * Tc * Tc * Tc);
+  // Eq. 1.5 page 15 of NaK handbook
+  const Real v_Na = 1. / (0.9453 - 2.2473e-4 * Tc);
+  drho_dT =
+      1000 *
+      -(_Nk * -(-2.172e-4 - 2 * 2.7e-8 * Tc + 3 * 4.77e-12 * Tc * Tc) * MathUtils::pow(v_k, 2) +
+        (1 - _Nk) * 2.2473e-4 * MathUtils::pow(v_Na, 2)) /
+      MathUtils::pow(_Nk * v_k + (1 - _Nk) * v_Na, 2);
```
Look at you doing calculus

GiudGiud committed:
Address Alex's review
- group function definitions

Update modules/fluid_properties/doc/content/source/userobjects/NaKFluidProperties.md

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update modules/fluid_properties/include/userobjects/NaKFluidProperties.h

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

lindsayad commented on modules/fluid_properties/include/userobjects/NaKFluidProperties.h:
```diff
@@ -0,0 +1,80 @@
+/**
+ * NaK fluid properties as a function of pressure (Pa) and temperature (K).
+ *
+ * Origin of fluid properties:
+ * 1) density; thermal diffusivity; dynamic viscosity; specific heat capacity
+ *
+ * SODIUM-NaK ENGINEERING HANDBOOK Volume I
+ * Sodium Chemistry and Physical Properties
+ * O. J. FOUST, Editor
+ * ISBN 0 677 03020 4
```
```suggestion
 * Source for fluid properties
 * - density; thermal diffusivity; dynamic viscosity; specific heat capacity:
 * SODIUM-NaK ENGINEERING HANDBOOK Volume I
 * Sodium Chemistry and Physical Properties
 * O. J. FOUST, Editor
 * ISBN 0 677 03020 4
```

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1544379/) on 05ad3b9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24452/site)

This comment will be updated on new commits.

lindsayad commented on modules/fluid_properties/doc/content/source/userobjects/NaKFluidProperties.md:
```diff
@@ -0,0 +1,55 @@
+| :----------------------------- | :------------- | :------------- |
+| Density liquid Na, $\rho_{Na}$ (kg/m^3)       | $0.9453 - 2.2473e-4 * T_c$ ($T_c$ in C) | Equation 1.5 |
+| Density liquid K, $\rho_K$ (kg/m^3)       | $0.8415 - 2.172e-4 * T_c - 2.7e-8 * T_c^2 + 4.77e-12 * T_c^3$ ($T_c$ in C) | Equation 1.8  |
+| Density liquid NaK, $\rho$ (kg/m^3)       | 1 / (N_K / \rho_K + N_{Na} / \rho_{Na}) | Equation 1.9 |
```
```suggestion
| Density liquid NaK, $\rho$ (kg/m^3)       | $1 / (N_K / \rho_K + N_{Na} / \rho_{Na})$ | Equation 1.9 |
```

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1544420/) on 05ad3b9 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>395ab6</th>
      <th colspan=4>#24452 05ad3b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.36%</td>
      <td>85.36%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88275</td>
      <td>88275</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15139</td>
      <td>15139</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24452/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24452/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>395ab6</th>
      <th colspan=4>#24452 05ad3b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.09%</td>
      <td>85.17%</td>
      <td>+0.08%</td>
      <td>90.76%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6860</td>
      <td>6968</td>
      <td>+108</td>
      <td>108</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1202</td>
      <td>1213</td>
      <td>+11</td>
      <td>11</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24452/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24452/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24452/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
Thanks for the prompt review!
